File: web/src/components/bandPlan.ts
import type {
  BandAllocation,
  BandBlock,
  BandLane,
  BandPlan,
  BandService,
  ChannelParams,
} from "../lib/types";

export interface BandSpan {
  block: BandBlock;
  allocation: BandAllocation;
  left: number;
  width: number;
  startsInside: boolean;
  endsInside: boolean;
}

export interface BandMatch {
  laneId: string;
  laneName: string;
  allocation: BandAllocation;
}

export interface BandIdentity {
  laneId: string;
  laneName: string;
  block: BandBlock;
  allocation: BandAllocation;
  covered: BandAllocation[];
}

export function spansIn(
  plan: BandPlan,
  lane: BandLane,
  lowHz: number,
  visibleHz: number,
): BandSpan[] {
  if (!(visibleHz > 0)) {
    return [];
  }
  const highHz = lowHz + visibleHz;
  const spans: BandSpan[] = [];
  for (const block of lane.blocks) {
    if (block.stop_hz <= lowHz || block.start_hz >= highHz) {
      continue;
    }
    const allocation = plan.allocations[block.of];
    if (allocation === undefined) {
      continue;
    }
    const left = (block.start_hz - lowHz) / visibleHz;
    const right = (block.stop_hz - lowHz) / visibleHz;
    spans.push({
      block,
      allocation,
      left: Math.max(0, left),
      width: Math.min(1, right) - Math.max(0, left),
      startsInside: left >= 0,
      endsInside: right <= 1,
    });
  }
  return spans;
}

export function identify(plan: BandPlan, hz: number): BandIdentity[] {
  const found: BandIdentity[] = [];
  for (const lane of plan.lanes) {
    const block = lane.blocks.find((entry) => entry.start_hz <= hz && entry.stop_hz > hz);
    const allocation = block === undefined ? undefined : plan.allocations[block.of];
    if (block === undefined || allocation === undefined) {
      continue;
    }
    found.push({
      laneId: lane.id,
      laneName: lane.name,
      block,
      allocation,
      covered: (block.covered ?? [])
        .map((at) => plan.allocations[at])
        .filter((entry): entry is BandAllocation => entry !== undefined),
    });
  }
  return found;
}

export function provisionText(plan: BandPlan, layer: string, id: string): string | null {
  return plan.provisions?.find((found) => found.layer === layer && found.id === id)?.text ?? null;
}

export function suggestedAt(found: readonly BandIdentity[]): ChannelParams | null {
  let suggested: ChannelParams | null = null;
  for (const entry of found) {
    suggested = entry.allocation.suggested ?? suggested;
  }
  return suggested;
}

export function searchPlan(plan: BandPlan, query: string, limit = 40): BandMatch[] {
  const words = query
    .toLowerCase()
    .split(/[\s,]+/)
    .filter((word) => word.length >= 2);
  const hz = parseFrequency(query);
  if (words.length === 0 && hz === null) {
    return [];
  }

  const lanes = new Map<string, string>();
  for (const lane of plan.lanes) {
    for (const block of lane.blocks) {
      for (const at of [block.of, ...(block.covered ?? [])]) {
        const allocation = plan.allocations[at];
        if (allocation !== undefined && !lanes.has(allocation.id)) {
          lanes.set(allocation.id, lane.id === "allocation" ? "" : lane.name);
        }
      }
    }
  }

  const scored: { match: BandMatch; score: number; width: number }[] = [];
  for (const allocation of plan.allocations) {
    const covers = hz !== null && allocation.start_hz <= hz && allocation.stop_hz > hz;
    const haystack = haystackOf(allocation);
    const matched = words.filter((word) => haystack.includes(word)).length;
    if (!covers && matched === 0) {
      continue;
    }
    scored.push({
      score: (covers ? 100 : 0) + matched,
      width: allocation.stop_hz - allocation.start_hz,
      match: {
        laneId: allocation.layer,
        laneName: lanes.get(allocation.id) ?? "",
        allocation,
      },
    });
  }
  scored.sort((a, b) => b.score - a.score || a.width - b.width);
  return scored.slice(0, limit).map((entry) => entry.match);
}

export function parseFrequency(query: string): number | null {
  const match = /^\s*(\d+(?:[.,]\d+)?)\s*(ghz|mhz|khz|hz)?\s*$/i.exec(query);
  if (match?.[1] === undefined) {
    return null;
  }
  const value = Number.parseFloat(match[1].replace(",", "."));
  if (!Number.isFinite(value)) {
    return null;
  }
  switch (match[2]?.toLowerCase()) {
    case "ghz":
      return value * 1e9;
    case "khz":
      return value * 1e3;
    case "hz":
      return value;
    default:
      return value * 1e6;
  }
}

const FILL: Record<BandService, string> = {
  amateur: "bg-band-amateur/25",
  broadcast: "bg-band-broadcast/25",
  aeronautical: "bg-band-aeronautical/25",
  maritime: "bg-band-maritime/25",
  mobile: "bg-band-mobile/25",
  satellite: "bg-band-satellite/25",
  navigation: "bg-band-navigation/25",
  science: "bg-band-science/25",
  ism: "bg-band-ism/25",
  other: "bg-band-other/25",
};

const EDGE: Record<BandService, string> = {
  amateur: "bg-band-amateur",
  broadcast: "bg-band-broadcast",
  aeronautical: "bg-band-aeronautical",
  maritime: "bg-band-maritime",
  mobile: "bg-band-mobile",
  satellite: "bg-band-satellite",
  navigation: "bg-band-navigation",
  science: "bg-band-science",
  ism: "bg-band-ism",
  other: "bg-band-other",
};

export function serviceFill(service: BandService): string {
  return FILL[service];
}

export function serviceEdge(service: BandService): string {
  return EDGE[service];
}

export function serviceLabel(service: BandService): string {
  return service === "ism" ? "ISM" : service.charAt(0).toUpperCase() + service.slice(1);
}

function haystackOf(allocation: BandAllocation): string {
  const service = allocation.service === "amateur" ? "amateur ham" : allocation.service;
  return `${allocation.name} ${(allocation.aliases ?? []).join(" ")} ${service}`.toLowerCase();
}

File: apps/native/src/shell/band_plan.rs
use sdrmm_wire::bandplan::{BandAllocation, BandPlan, BandService};

#[derive(Clone, Debug, PartialEq)]
pub struct BandMatch {
    pub lane_id: String,
    pub lane_name: String,
    pub allocation: BandAllocation,
}

pub(crate) fn allocation_at(plan: &BandPlan, at: u32) -> Option<&BandAllocation> {
    plan.allocations.get(usize::try_from(at).ok()?)
}

#[must_use]
pub fn band_tune_hz(allocation: &BandAllocation) -> f64 {
    let middle = allocation.start_hz + (allocation.stop_hz - allocation.start_hz) / 2.0;
    match allocation.channel_step_hz.filter(|step| *step > 0.0) {
        Some(step) => allocation.start_hz + ((middle - allocation.start_hz) / step).round() * step,
        None => middle,
    }
}

fn haystack_of(allocation: &BandAllocation) -> String {
    let service = if allocation.service == BandService::Amateur {
        "amateur ham"
    } else {
        service_token(allocation.service)
    };
    format!(
        "{} {} {service}",
        allocation.name,
        allocation.aliases.join(" ")
    )
    .to_lowercase()
}

fn lane_names(plan: &BandPlan) -> Vec<(String, String)> {
    let mut lanes: Vec<(String, String)> = Vec::new();
    for lane in &plan.lanes {
        for block in &lane.blocks {
            for at in std::iter::once(block.of).chain(block.covered.iter().copied()) {
                let Some(allocation) = allocation_at(plan, at) else {
                    continue;
                };
                if !lanes.iter().any(|(id, _)| *id == allocation.id) {
                    let name = if lane.id == "allocation" {
                        String::new()
                    } else {
                        lane.name.clone()
                    };
                    lanes.push((allocation.id.clone(), name));
                }
            }
        }
    }
    lanes
}

#[must_use]
pub fn search_plan(plan: &BandPlan, query: &str, limit: usize) -> Vec<BandMatch> {
    let lowered = query.to_lowercase();
    let words: Vec<&str> = lowered
        .split(|ch: char| ch.is_whitespace() || ch == ',')
        .filter(|word| word.chars().count() >= 2)
        .collect();
    let hz = parse_frequency(query);
    if words.is_empty() && hz.is_none() {
        return Vec::new();
    }
    let lanes = lane_names(plan);
    let mut scored: Vec<(u32, f64, BandMatch)> = plan
        .allocations
        .iter()
        .filter_map(|allocation| {
            let covers = hz.is_some_and(|hz| allocation.start_hz <= hz && allocation.stop_hz > hz);
            let haystack = haystack_of(allocation);
            let matched = words
                .iter()
                .filter(|word| haystack.contains(**word))
                .count();
            if !covers && matched == 0 {
                return None;
            }
            let lane_name = lanes
                .iter()
                .find(|(id, _)| *id == allocation.id)
                .map(|(_, name)| name.clone())
                .unwrap_or_default();
            Some((
                u32::from(covers) * 100 + u32::try_from(matched).unwrap_or(u32::MAX),
                allocation.stop_hz - allocation.start_hz,
                BandMatch {
                    lane_id: allocation.layer.clone(),
                    lane_name,
                    allocation: allocation.clone(),
                },
            ))
        })
        .collect();
    scored.sort_by(|a, b| b.0.cmp(&a.0).then_with(|| a.1.total_cmp(&b.1)));
    scored
        .into_iter()
        .take(limit)
        .map(|(_, _, found)| found)
        .collect()
}

#[must_use]
pub fn parse_frequency(query: &str) -> Option<f64> {
    let trimmed = query.trim();
    let split = trimmed
        .find(|ch: char| !(ch.is_ascii_digit() || ch == '.' || ch == ','))
        .unwrap_or(trimmed.len());
    let (number, unit) = trimmed.split_at(split);
    let valid_number = {
        let mut parts = number.splitn(2, ['.', ',']);
        let whole = parts.next().unwrap_or_default();
        let fraction = parts.next();
        !whole.is_empty()
            && whole.chars().all(|ch| ch.is_ascii_digit())
            && fraction.is_none_or(|fraction| {
                !fraction.is_empty() && fraction.chars().all(|ch| ch.is_ascii_digit())
            })
    };
    if !valid_number {
        return None;
    }
    let value: f64 = number.replace(',', ".").parse().ok()?;
    let scale = match unit.trim().to_lowercase().as_str() {
        "ghz" => 1e9,
        "mhz" | "" => 1e6,
        "khz" => 1e3,
        "hz" => 1.0,
        _ => return None,
    };
    value.is_finite().then_some(value * scale)
}

#[must_use]
pub fn service_label(service: BandService) -> &'static str {
    match service {
        BandService::Amateur => "Amateur",
        BandService::Broadcast => "Broadcast",
        BandService::Aeronautical => "Aeronautical",
        BandService::Maritime => "Maritime",
        BandService::Mobile => "Mobile",
        BandService::Satellite => "Satellite",
        BandService::Navigation => "Navigation",
        BandService::Science => "Science",
        BandService::Ism => "ISM",
        BandService::Other => "Other",
    }
}

#[must_use]
pub fn service_token(service: BandService) -> &'static str {
    match service {
        BandService::Amateur => "amateur",
        BandService::Broadcast => "broadcast",
        BandService::Aeronautical => "aeronautical",
        BandService::Maritime => "maritime",
        BandService::Mobile => "mobile",
        BandService::Satellite => "satellite",
        BandService::Navigation => "navigation",
        BandService::Science => "science",
        BandService::Ism => "ism",
        BandService::Other => "other",
    }
}

#[allow(dead_code)]
pub mod ruler;

#[cfg(test)]
mod tests;
